### PR TITLE
[FIX] Add data access for containers

### PR DESCRIPTION
**Pull request by eseiler**
* Adds `raw_data()` for `bitcompressed_vector` and allows access of the underlying SDSL vector.
* Renames `data()` to `raw_data()` in `concatenated_sequences` since `data()` implies contiguous memory locations.
* `small_vector` and `small_string` have `data()` and are contiguous, ergo no changes here.

### CODE REVIEW

**Comment by svnbgnk** [comment]
The operator compare all the bitcompressed data from one vector to another, right?
https://github.com/seqan/seqan3/blob/25d06a0aeb0e775a9c8c44ca8c93848aef59a127/include/seqan3/range/container/bitcompressed_vector.hpp#L978

File: test/unit/range/container/container_of_container_test.cpp [hunk 1]
```diff
@@ -152,10 +152,10 @@ TYPED_TEST(container_of_container, element_access)
         EXPECT_EQ(dna4_vector(t2.concat()), "ACGTACGTGAGGA"_dna4);
 
         // data
-        EXPECT_EQ(std::get<0>(t1.data()), "ACGTACGTGAGGA"_dna4);
-        EXPECT_EQ(std::get<0>(t2.data()), "ACGTACGTGAGGA"_dna4);
-        EXPECT_EQ(std::get<1>(t1.data()), (std::vector<size_type>{0, 4, 8, 13}));
-        EXPECT_EQ(std::get<1>(t2.data()), (std::vector<size_type>{0, 4, 8, 13}));
+        EXPECT_EQ(std::get<0>(t1.raw_data()), "ACGTACGTGAGGA"_dna4);
```
Comment (svnbgnk) [inline test/unit/range/container/container_of_container_test.cpp]: t1.raw_data() return a pair. I did not know that you can also use std::get(). IMO this is a bit confusing.

**Comment by eseiler** [comment]
> The operators compare all the bitcompressed data from one vector to another, right?
> https://github.com/seqan/seqan3/blob/25d06a0aeb0e775a9c8c44ca8c93848aef59a127/include/seqan3/range/container/bitcompressed_vector.hpp#L978-L981

We delegate to the compairson operator of the underlying data type (sdsl::int_vector), which then compares the content.

**Comment by eseiler** [comment]
> One thing I noticed that did not data with raw_data() swap:
> https://github.com/seqan/seqan3/blob/25d06a0aeb0e775a9c8c44ca8c93848aef59a127/include/seqan3/range/container/bitcompressed_vector.hpp#L937-L941
> 
> .
> Is there any reason for that?

No reason, just didn't see it. I can access the private `data` member within the class, but I'll change it for consistency.

File: test/unit/range/container/container_of_container_test.cpp (hunk 1 above)
Comment (eseiler) [reply test/unit/range/container/container_of_container_test.cpp]: `std::get` is usually how you access data of pair/tuple.
Also see https://en.cppreference.com/w/cpp/utility/pair

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 2]
```diff
@@ -518,6 +518,19 @@ class bitcompressed_vector
         return (*this)[size()-1];
     }
 
+    //!\brief Direct access to the underlying SDSL vector.
```
Comment (h-2) [inline include/seqan3/range/container/bitcompressed_vector.hpp]: Maybe you should describe more clearly what the type will be as this is exposed to the API. Note that if we expose this to the API, we cannot change the internal type anymore without breakage.

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 3]
```diff
@@ -964,37 +977,37 @@ class bitcompressed_vector
     //!\brief Checks whether `*this` is equal to `rhs`.
     constexpr bool operator==(bitcompressed_vector const & rhs) const noexcept
     {
-        return data == rhs.data;
+        return raw_data() == rhs.raw_data();
```
Comment (h-2) [inline include/seqan3/range/container/bitcompressed_vector.hpp]: Are these changes necessary? I don't care much ....

File: include/seqan3/range/container/concatenated_sequences.hpp [hunk 4]
```diff
@@ -602,16 +602,30 @@ class concatenated_sequences
...
+    {
+        return raw_data();
+    }
+
+    //!\copydoc raw_data()
+    //!\deprecated Use raw_data() instead.
+    SEQAN3_DEPRECATED_310 std::pair<decltype(data_values) const &, decltype(data_delimiters) const &> data() const
+    {
+        return raw_data();
+    }
```
Comment (h-2) [inline include/seqan3/range/container/concatenated_sequences.hpp]: :+1:

File: include/seqan3/range/container/bitcompressed_vector.hpp (hunk 2 above)
Comment (eseiler) [reply include/seqan3/range/container/bitcompressed_vector.hpp]: I changed it a bit. I'd like to link to a documentation of the bitvector. But there are no online docs for the sdsl3 bitvector. And the sdsl2 one looks like [this](https://algo2.iti.kit.edu/gog/docs/html/namespacesdsl.html#a93df8f263fced9ae9a8b0f9f37e444eb).

Anyway, I noticed that in `concatenated_sequences` we added this sentence:
`This exact representation of the data is implementation defined. Do not rely on it for API stability.`
So it's basically not API and we could add `\noapi` for doxygen to this function?

Then we could discuss if all our container `data`/`raw_data` functions are no api.

File: include/seqan3/range/container/bitcompressed_vector.hpp (hunk 3 above)
Comment (eseiler) [reply include/seqan3/range/container/bitcompressed_vector.hpp]: Mhm, I was afraid of the code coverage since I didn't want to test the `data`/`raw_data` functions.

I'll revert it and see what happens.

File: include/seqan3/range/container/bitcompressed_vector.hpp (hunk 2 above)
Comment (h-2) [reply include/seqan3/range/container/bitcompressed_vector.hpp]: > So it's basically not API and we could add \noapi for doxygen to this function?

:+1:

> Then we could discuss if all our container data/raw_data functions are no api.

:+1:

File: CHANGELOG.md [hunk 5]
```diff
@@ -40,6 +40,10 @@ If possible, provide tooling that performs the changes, e.g. a shell-script.
 * **The `type_list` header has moved:**
   If you included `<seqan3/core/type_list.hpp>` you need to change the path to `<seqan3/core/type_list/type_list.hpp>`.
 
+### The `data()` function of `seqan3::concatenated_sequences` has been deprecated
+
+Use `raw_data()` instead.
```
[the remark this answers is not in the document]
Comment (eseiler) [reply CHANGELOG.md]: If we mark the functions as noapi we don't need to include it in the changelog?
